G-API: oneVPL (simplification) added CPU, DX11(fallback CPU) accels & surface pool

sivanov-work opened the pull request:
this PR is the single one in series of https://github.com/opencv/opencv/pull/20469
Added VPL CPU accelerator, DX11 ( fallback CPU) accelerator & surface pool

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### Build configuration

```
force_builders=Custom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2021.3.0:20.04
build_image:Custom Win=openvino-2021.2.0
build_image:Custom Mac=openvino-2021.2.0

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.cpp:
```diff
@@ -0,0 +1,227 @@
+
+        });
+    size_t i = 0;
+    try {
+        for (; i < pool_size; i++) {
+            size_t preallocated_mem_offset = static_cast<size_t>(i) * surface_size_bytes;
+
+            surface_ptr_t surf_ptr = creator(workspace_mem_owner,
+                                             preallocated_mem_offset,
+                                             preallocated_raw_bytes);
+            pool.push_back(std::move(surf_ptr));
+        }
+    } catch (const std::exception& ex) {
+        throw std::runtime_error(std::string("VPLCPUAccelerationPolicy::create_surface_pool - ") +
+                                 "cannot construct surface index: " + std::to_string(i) + ", error: " +
+                                 ex.what() +
+                                 "Requested surface count: " + std::to_string(pool_size) +
+                                 ", surface bytes: " + std::to_string(surface_size_bytes));
+    }
```
Do we really need this handling here? What throws here in particular - is it our code or the external one?

Any exception handling within our module complicates the future debug (or I miss something)

dmatveev replied:
ok let's see in the future PRs

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.cpp:
```diff
@@ -0,0 +1,227 @@
+    GAPI_LOG_INFO(nullptr, "New pool allocated, key: " << preallocated_pool_memory_ptr <<
+                           ", surface count: " << pool.size() <<
+                           ", surface size bytes: " << surface_size_bytes);
+    try {
+        if (!pool_table.emplace(preallocated_pool_memory_ptr, std::move(pool)).second) {
+            throw std::runtime_error(std::string("VPLCPUAccelerationPolicy::create_surface_pool - ") +
+                                     "cannot insert pool, table size: " + std::to_string(pool_table.size()));
+        }
+    } catch (const std::exception&) {
+        throw;
+    }
```
Don't think it is required here too. Probably a meaningful log line would help more instead

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.cpp:
```diff
@@ -0,0 +1,227 @@
+        std::stringstream ss;
+        ss << "key is not found: " << key << ", table size: " << pool_table.size();
+        const std::string& str = ss.str();
+        GAPI_LOG_WARNING(nullptr, str);
```
I may be wrong but LOG macros were able to support `<<` formatting

dmatveev replied:
Maybe I confuse this with some other project already :)

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.cpp:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    pool_t& requested_pool = pool_it->second;
+#ifdef TEST_PERF
```
Where is this definition set? should it be `OPENCV_GAPI_` something?

dmatveev replied:
ok then please put a comment about this macro somewhere in the code so this will be removed

dmatveev replied:
(because usually it just stays like this :) )

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.cpp:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    return cv::MediaFrame::AdapterPtr{new VPLMediaFrameCPUAdapter(*it)};
+#endif // TEST_PERF
```
y do we need this everywhere?

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.hpp:
```diff
@@ -0,0 +1,55 @@
+
+struct VPLCPUAccelerationPolicy final : public VPLAccelerationPolicy
+{
+    // GAPI_EXPORTS for tests
+    GAPI_EXPORTS VPLCPUAccelerationPolicy();
+    GAPI_EXPORTS ~VPLCPUAccelerationPolicy();
+#ifdef TEST_PERF
+    using pool_t = CachedPool;
+#else  // TEST_PERF
+    using pool_t = std::vector<surface_ptr_t>;
+#endif // TEST_PERF
+
+    GAPI_EXPORTS void init(session_t session) override;
+    GAPI_EXPORTS void deinit(session_t session) override;
+    GAPI_EXPORTS pool_key_t create_surface_pool(size_t pool_size, size_t surface_size_bytes, surface_ptr_ctr_t creator) override;
+    GAPI_EXPORTS surface_weak_ptr_t get_free_surface(pool_key_t key) override;
+    GAPI_EXPORTS size_t get_free_surface_count(pool_key_t key) const override;
+    GAPI_EXPORTS size_t get_surface_count(pool_key_t key) const override;
```
I'd export the whole class and that's it

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.cpp:
```diff
@@ -0,0 +1,114 @@
+
+#ifdef HAVE_DIRECTX
+#ifdef HAVE_D3D11
+#pragma comment(lib,"d3d11.lib")
```
shouldn't this be handled by cmake? may this lead to some unexpected behavior? cc: @alalek

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.cpp:
```diff
@@ -0,0 +1,114 @@
+namespace wip {
+VPLDX11AccelerationPolicy::VPLDX11AccelerationPolicy()
+{
+#ifdef CPU_ACCEL_ADAPTER
+    adapter.reset(new VPLCPUAccelerationPolicy);
+#endif
```
what is this `#ifdef` and why it is a special case?

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.hpp:
```diff
@@ -0,0 +1,66 @@
+#define GAPI_STREAMING_ONEVPL_ACCELERATORS_ACCEL_POLICY_DX11_HPP
+
+#include "opencv2/gapi/own/exports.hpp" // GAPI_EXPORTS
+//TODO Remove the next MACRO right after DX11 implementation
+#define  CPU_ACCEL_ADAPTER
```
But why?

dmatveev replied:
ok, let's avoid this macro-defined behavior in the future (if I get what's being addressed there)

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.hpp:
```diff
@@ -0,0 +1,66 @@
+namespace gapi {
+namespace wip {
+
+struct VPLDX11AccelerationPolicy final: public VPLAccelerationPolicy
```
Again, I'd export the whole structure

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/surface/surface_pool.cpp:
```diff
@@ -0,0 +1,78 @@
+                return !val->get_locks_count();
+            });
+
+        // Limitation realloc pool might be a future extension
```
Don't get what "limitation realloc pool" is. Is the pool a feature or limitation?

dmatveev replied:
ok I agree

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/surface/surface_pool.cpp:
```diff
@@ -0,0 +1,78 @@
+}
+
+CachedPool::surface_ptr_t CachedPool::find_by_handle(mfxFrameSurface1* handle) {
+    auto it = cache.find(handle);
+    if (it == cache.end()) {
+        std::stringstream ss;
+        ss << "cannot get requested surface from pool, surf: "
+           << handle << ", pool size: " << surfaces.size();
+        const std::string& str = ss.str();
+        GAPI_LOG_WARNING(nullptr, str);
+        throw std::runtime_error(std::string(__FUNCTION__) + " - " + str);
+    }
+    return it->second;
```
I don't think we need that extensive error hanlding for a simple `map.find()` line

sivanov-work replied:
i agree, looks like on data corruption: Assert must be here

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/surface/surface_pool.cpp:
```diff
@@ -0,0 +1,78 @@
+    next_free_it = surfaces.begin();
+}
+
+CachedPool::surface_ptr_t CachedPool::find_free() {
```
Is this pool supposed to be thread safe or the protection is done outside?

dmatveev replied:
MediaFrames may be destroyed concurrently, is it ok for this implementation?

sivanov-work replied:
yes, it is ok. It just decreased surface shared counter as `shared_ptr<surface>` and ApplicationCounter inside `surface`

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/surface/surface_pool.hpp:
```diff
@@ -0,0 +1,45 @@
+namespace wip {
+
+class Surface;
+class CachedPool {
```
Just export the whole class

sivanov-work replied:
ok

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/surface/surface_pool.cpp:
```diff
@@ -0,0 +1,78 @@
+                            [](const surface_ptr_t& val) {
+                    GAPI_DbgAssert(val && "Pool contains empty surface");
+                    return !val->get_locks_count();
+                });
```
circular buffer imitation

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.cpp:
```diff
@@ -0,0 +1,227 @@
+        pair.second.clear();
+        // do not free key here: last surface will release it
+    }
+    pool_table.clear();
```
> So clear for pool_table here doesn't means that all pointers released until last consumer takes MediaFrame.

Sorry but all this object lifetime thing is still not very clear to me. Can you please explain in detail how these different objects relate?

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.cpp:
```diff
@@ -0,0 +1,227 @@
+#endif
+
+    if (!preallocated_pool_memory_ptr) {
+        throw std::runtime_error("VPLCPUAccelerationPolicy::create_surface_pool - failed: not enough memory."
```
+1 to the second one, and I am definitely sure we'll have to keep Linux in mind - maybe in the backlog

sivanov-work replied:
Issue was added in backlog

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_cpu.cpp:
```diff
@@ -0,0 +1,227 @@
+        GAPI_LOG_INFO(nullptr, "Released workspace memory: " << ptr);
+        ptr = nullptr;
+#else
+        abort(); //not implemented
```
ok

dmatveev commented on modules/gapi/src/streaming/onevpl/accelerators/surface/surface_pool.cpp:
```diff
@@ -0,0 +1,71 @@
+            }
+        }
+
+        next_free_it = it;
+        ++next_free_it;
+
+    return *it;
```
Formatting may look strange

dmatveev reviewed (approved):
ok to me, stuff can be fixed with the next set of PRs

sivanov-work commented:
@alalek  or @mshabunin  could you please merge it?
There is an one unaddressed comment about DX11 linkage: https://github.com/opencv/opencv/pull/20727/files#r713648649
Is there any best known practices from your side?
